palace/fem/mesh.cpp

```diff
@@ -0,0 +1,369 @@
+    {
+      if (const auto *submesh = dynamic_cast<const mfem::ParSubMesh *>(&mesh))
+      {
+        MFEM_VERIFY(submesh->GetFrom() == mfem::SubMesh::From::Boundary,
+                    "Unexpected non-SubMesh object for BuildCeedGeomFactorData with Mesh "
+                    "with (dim, space_dim) = ("
+                        << mesh.Dimension() << ", " << mesh.SpaceDimension() << ")!");
+        return [&, submesh](int i)
+        {
+          // Mesh is actually a boundary submesh, so we use the boundary attribute mappings
+          // from the parent mesh.
+          const int attr = mesh.GetAttribute(i);
+          const int nbr_attr = GetBdrNeighborAttribute(submesh->GetParentElementIDMap()[i],
+                                                       *submesh->GetParent(), FET, T1, T2);
+          MFEM_ASSERT(loc_bdr_attr.find(attr) != loc_bdr_attr.end() &&
+                          loc_bdr_attr.at(attr).find(nbr_attr) !=
+                              loc_bdr_attr.at(attr).end(),
+                      "Missing libCEED boundary attribute for attribute " << attr << "!");
+          return loc_bdr_attr.at(attr).at(nbr_attr);
+        };
+      }
+      else
+      {
+        return [&](int i)
+        {
+          const int attr = mesh.GetAttribute(i);
+          MFEM_ASSERT(loc_attr.find(attr) != loc_attr.end(),
+                      "Missing libCEED domain attribute for attribute " << attr << "!");
+          return loc_attr.at(attr);
+        };
+      }
+    }();
+    for (auto &[geom, indices] : element_indices)
+    {
+      ceed::CeedGeomFactorData data =
+          AssembleGeometryData(*mesh.GetNodes(), ceed, geom, indices, GetCeedAttribute);
+      geom_data_map.emplace(geom, std::move(data));
+    }
+  }
+
+  // Then boundary elements (no support for boundary integrators on meshes embedded in
+  // higher dimensional space for now).
+  if (mesh.Dimension() == mesh.SpaceDimension())
+  {
+    const int nbe = mesh.GetNBE();
+    const int stride = (nbe + nt - 1) / nt;
+    const int start = i * stride;
+    const int stop = std::min(start + stride, nbe);
+    constexpr bool use_bdr = true;
+    auto element_indices = GetElementIndices(mesh, use_bdr, start, stop);
+    auto GetCeedAttribute = [&](int i)
+    {
+      const int attr = mesh.GetBdrAttribute(i);
+      const int nbr_attr = GetBdrNeighborAttribute(i, mesh, FET, T1, T2);
+      MFEM_ASSERT(loc_bdr_attr.find(attr) != loc_bdr_attr.end() &&
+                      loc_bdr_attr.at(attr).find(nbr_attr) != loc_bdr_attr.at(attr).end(),
+                  "Missing libCEED boundary attribute for attribute " << attr << "!");
+      return loc_bdr_attr.at(attr).at(nbr_attr);
+    };
+    for (auto &[geom, indices] : element_indices)
+    {
+      ceed::CeedGeomFactorData data =
+          AssembleGeometryData(*mesh.GetNodes(), ceed, geom, indices, GetCeedAttribute);
+      geom_data_map.emplace(geom, std::move(data));
+    }
+  }
```

Comment:
I don't think this is what you want here, it's fine to assemble the domain geometry data for an embedded mesh and this `MFEM_ABORT` would not allow that. I do agree that it would be nice if someone added a boundary integrator to an embedded mesh they get an error message, but I think it will come later (from `BilinearForm`, the integrator would not be assembled). This is unit-tested so I'm not too worried about it.